src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientKeyedLifetime.cs

```diff
@@ -0,0 +1,92 @@
+namespace Microsoft.Extensions.DependencyInjection
+{
+    internal class HttpClientKeyedLifetime
+    {
+        public static readonly HttpClientKeyedLifetime Disabled = new(null!, null!, null!);
+
+        public object ServiceKey { get; }
+        public ServiceDescriptor Client { get; }
+        public ServiceDescriptor Handler { get; }
+
+        public bool IsDisabled => ReferenceEquals(this, Disabled);
+
+        private HttpClientKeyedLifetime(object serviceKey, ServiceDescriptor client, ServiceDescriptor handler)
+        {
+            ServiceKey = serviceKey;
+            Client = client;
+            Handler = handler;
+        }
+
+        private HttpClientKeyedLifetime(object serviceKey, ServiceLifetime lifetime)
+        {
+            ThrowHelper.ThrowIfNull(serviceKey);
+            ServiceKey = serviceKey;
+            Client = ServiceDescriptor.DescribeKeyed(typeof(HttpClient), ServiceKey, CreateKeyedClient, lifetime);
+            Handler = ServiceDescriptor.DescribeKeyed(typeof(HttpMessageHandler), ServiceKey, CreateKeyedHandler, lifetime);
+        }
+
+        public HttpClientKeyedLifetime(ServiceLifetime lifetime) : this(KeyedService.AnyKey, lifetime) { }
+        public HttpClientKeyedLifetime(string name, ServiceLifetime lifetime) : this((object)name, lifetime) { }
+
+        public void AddRegistration(IServiceCollection services)
+        {
+            if (IsDisabled)
+            {
+                return;
+            }
+
+            services.Add(Client);
+            services.Add(Handler);
+        }
+
+        public void RemoveRegistration(IServiceCollection services)
+        {
+            if (IsDisabled)
+            {
+                return;
+            }
+
+            services.Remove(Client);
+            services.Remove(Handler);
+        }
+
+        private static HttpClient CreateKeyedClient(IServiceProvider serviceProvider, object? key)
+        {
+            if (key is not string name || IsKeyedLifetimeDisabled(serviceProvider, name))
+            {
+                return null!;
+            }
+            return serviceProvider.GetRequiredService<IHttpClientFactory>().CreateClient(name);
+        }
+
+        private static HttpMessageHandler CreateKeyedHandler(IServiceProvider serviceProvider, object? key)
+        {
+            if (key is not string name || IsKeyedLifetimeDisabled(serviceProvider, name))
+            {
+                return null!;
+            }
+            HttpMessageHandler handler = serviceProvider.GetRequiredService<IHttpMessageHandlerFactory>().CreateHandler(name);
+            // factory will return a cached instance, wrap it to be able to respect DI lifetimes
+            return new LifetimeTrackingHttpMessageHandler(handler);
+        }
+
+        private static bool IsKeyedLifetimeDisabled(IServiceProvider serviceProvider, string name)
+        {
+            HttpClientMappingRegistry registry = serviceProvider.GetRequiredService<HttpClientMappingRegistry>();
```

Comment:
If `serviceProvider.GetRequiredService` is expensive, please consider, for perf reasons, store/cache HttpClientMappingRegistry

Reply:
> If serviceProvider.GetRequiredService is expensive

I don't think so, the registry is a singleton with an implementation instance provided right from the start, it should be super fast :D

Also there's nowhere to store it to anyway, it's a static method that is used as a part of a factory method in DI for the keyed registrations.

it's like, the registry _is_ the cache here for all the mapping information

Reply:
Great to hear our DI is efficient. I have some performance battles scars from time I have used Ninject where getting service was not cheap at all :-)

Reply:
The service provider is pretty fast at resolving singletons since it's basically a concurrent dictionary lookup, but it wouldn't be nearly as fast as returning the `IHttpClientFactory` and `IHttpMessageHandlerFactory` cached in a field of `HttpClientKeyedLifetime`. This is particularly true when there are multiple `GetService` calls involved as is the case here.

The downside of caching like this is that it could return a service from the wrong provider if someone builds multiple service providers from the same `IServiceCollection`. The expansion of `HttpClientMappingRegistry` to do things beyond validation already makes me nervous, because copying the `IServiceCollection` and then making modifications to one but not the other could lead to some very weird behavior, but at least that's pretty rare. Rebuilding the same `IServiceCollection` is relatively common though.

I agree we better off not caching to avoid potential breaks given multiple providers built from the same collection, but I'm curious whether we consider that a supported scenario. Do we have tests for that? What about copying the service collection and then making modifications to only one copy?